### ISSUE TITLE
Decoupling: Platform-agnostic media upload

### ISSUE THREAD

**Issue by swissspidy** [GoogleForCreators/web-stories-wp#8953]
## Background

The story editor package uses `useMediaPicker()` in multiple places to handle media uploads.

That hook allows using the WordPress media modal for upload.

To decouple the editor from WordPress, this logic needs to be made more platform-agnostic, with the WordPress media modal handling moved to `wp-story-editor`, allowing other implementations on other platforms.

## Proof-of-concept

@sayedtaqui has worked on a POC for decoupling the media upload, making `MediaUpload` a React component that needs to be passed via the `ConfigProvider`:

https://github.com/rtCamp/web-stories-wp/pull/731

## Context

Gutenberg uses the `@wordpress/hooks` package to allow filtering the `MediaUpload` component. By default it renders nothing, but in the WordPress block editor it's filtered to render a component using the WordPress media modal.

https://github.com/WordPress/gutenberg/blob/3e14c5e812d8077f28e080522f20763bbfb1dccc/packages/block-editor/src/components/media-upload/README.md

https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/block-editor/src/components/media-upload/index.js

https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/edit-post/src/hooks/components/index.js

We can't use `@wordpress/hooks` directly due to the GPL license, which is incompatible with our Apache license.

## Implementation Brief

We need to find a good way for making media upload platform-agnostic.

The POC above works, but the config React context is better suited for passing data/settings and not React components. So it doesn't feel right to pass `MediaUpload` that way.

There's also another POC for using hooks/filters: https://github.com/rtCamp/web-stories-wp/pull/765

But we can't really use that as-is due to licensing restrictions (uses parts of GPL-licensed code).

An alternative could perhaps be to use a [Slot & Fill technique](https://github.com/camwest/react-slot-fill).

But maybe there are other, better ways.

**Comment by swissspidy** [comment]
@choumx @barklund Would love to hear your thoughts on this

**Comment by swissspidy** [comment]
Current plan is to proceed with the POC and refine later based on feedback. The code needs to be moved anyway.

**Comment by samwhale** [comment]
Hi hi hi! So just opened #9534 and have been learning about this work that's being done. 

@sayedtaqui brought up this comment: https://github.com/google/web-stories-wp/pull/9534#discussion_r736658207 which is a good thing to consider.

I’m torn about how to make the "replace media" functionality with quick tips work (considering this work in flight and the current idea of the `MediaUpload` component. To make it work now we’d have to either:
1. add in code to the context menu to render wrappers around buttons (or just any UI) :grimacing:
2. write our own hook that does the exact same thing as useMediaPicker

---

**Option 1**

The `ContextMenu` is supposed to be purely presentational. List items are rendered by passing in objects with `onClick` event handlers. There currently is no way to render anything other than the UI the context renders by default.

We'd have to add a render prop there to render different UI, or a wrapper around the current UI. I think this is kind of icky, but do-able.

---

**Option 2**

We'd write a hook that's very similar to the one in the `wp-story-editor` package. It would do what we needed it to, but future updates to media uploading would have to be done in two places. That would be frustrating to try to remember and keep track of.

It would be ideal to only have _one_ way to upload media.

---

🤔 Would we not be able to just pass in the "upload media" endpoint through the config and do the UI + handling in the app? Do users need to have the ability to provide their own `MediaUpload` component?

**Comment by swissspidy** [comment]
> Do users need to have the ability to provide their own MediaUpload component?

Not _users_ (as in people), but platforms/consumers

> 🤔 Would we not be able to just pass in the "upload media" endpoint through the config and do the UI + handling in the app?

Let me give you an example why that doesn't work:

We want to be able to create a standalone story editor PWA that does not run on WordPress or anything. Just a plain simple React app basically.

In such a context you don't have any WordPress media modal or any "upload media" endpoint or anything. Instead, uploading files would mean you'd open a file picker to select a file from your computer.

So WordPress provides aa `MediaUpload` component that opens the WordPress media picker, whereas the PWA provides a component that opens the OS file picker.

This ticket is about discussing/defining the best way to provide that component. So the quick tips use case is very useful context for that discussion!

**Comment by samwhale** [comment]
Ah thanks for the context. I see now why that wouldn't work. 

💭 Maybe we get the component from the config and render it at the top level of the app - but the component opens based off of state internal to our app. That way we could pass control of the visibility to the rest of the app.

Opening the modal could also pass in a callback to call when selecting media. maybe something like this:

```js
const { onOpenModal } = useMediaContext();

...

const handleAddMediaClick = (ev) => {
    function insertMediaCallback() { ... }

    onOpenModal(ev, insertMediaCallback)
}
```

That way we can inject the functionality the app needs and get access to the resource (and the added benefit of only rendering the component once). However, looking at the `useMediaPicker` code maybe this wouldn't be as straightforward as the above.

**Comment by dreamofabear** [comment]
After a quick review, I think that injecting a `MediaUpload` React component via config sounds reasonable and idiomatic to React (and preferable to hooks or other APIs). 

> Maybe we get the component from the config and render it at the top level of the app

IMO we should avoid baking in any assumptions about the UX of the injected `MediaUpload`. For the WP use case (a modal popup), that should be handled by the WP-specific `MediaUpload` implementation internals. That said, lifecycle callbacks/events will probably be useful on the component API.

**Comment by samwhale** [comment]
Seems like the path forward for the replace media quick action would be to add a way for the context menu to render the `MediaUpload` component around/as the button in the menu. Although it's not ideal, I think that would make the most sense.

**Comment by maxyinger** [comment]
yea I think with the way the context menu is setup, the way we've decided to setup the `MediaPicker` api, and the dependency structure, this will be weird to implement.

I think you're either going to have to:
1) Augment the context menu items to take a render prop and pass in the `MediaPicker` to relevant items. This will most likely feel hacky but might be the path of least resistance
2) Create some wrapper technique around `MediaPicker` that renders it, but reaches in and exposes its `open` method, which you can then consume and pass into the context menu.

I'm not in love with either option, but those are the only two routes I see with the given `MediaPicker` spec and existing context menu setup.

I do agree that making the config spec take a hook instead of a component would be drastically better to use in this instance, but I'm assuming the component based spec has a lot of other benefits elsewhere (including lots of prior adoption).

I guess my only critique of the `MediaPicker` component is that it feels a little less in vogue to use a render prop called `render` when you can just use the `children` prop, but that's such a minor detail 🤷  ie:
```
function MediaUpload({ children, ...props }) {
  const open = useMediaPicker(props);
  return children({ open });
}
```

anyway, tldr; I think it will be messy, but I don't see any better options

**Comment by swissspidy** [comment]
Closing this one for now as we decided to continue with the status quo
